src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,396 @@
+     *         private static final MemoryLayout CAPTURE_LAYOUT =
+     *                 Linker.Option.captureStateLayout();
+     *         private static final BufferStack POOL =
+     *                 BufferStack.of(CAPTURE_LAYOUT);
+     *
+     *         public int invoke(MethodHandle target,
+     *                           String stateName,
+     *                           int a, int b) {
+     *             try (var arena = POOL.pushFrame(CAPTURE_LAYOUT)) {
+     *                 final MemorySegment segment = arena.allocate(CAPTURE_LAYOUT);
+     *                 final int result = (int) handle.invoke(segment, a, b);
+     *                 if (result >= 0) {
+     *                     return result;
+     *                 }
+     *                 return -(int) CAPTURE_LAYOUT
+     *                     .varHandle(MemoryLayout.PathElement.groupElement(stateName))
+     *                         .get(segment, 0);
+     *             }
+     *         }
+     *}
+     * except it is more performant. In the above {@code stateName} is the name of the
+     * captured state (e.g. {@code errno}). The static {@code CAPTURE_LAYOUT} is shared
+     * across all target method handles adapted by this method.
+     *
+     * @param target    method handle that returns an {@code int} or a {@code long} and
+     *                  has a capturing state MemorySegment as its first parameter
+     * @param stateName the name of the capturing state member layout (i.e. "errno",
+     *                  "GetLastError", or "WSAGetLastError")
+     * @throws IllegalArgumentException if the provided {@code target}'s return type is
+     *                                  not {@code int} or {@code long}
+     * @throws IllegalArgumentException if the provided {@code target}'s first parameter
+     *                                  type is not {@linkplain MemorySegment}
+     * @throws IllegalArgumentException if the provided {@code stateName} is unknown on
+     *                                  the current platform
+     */
+    public static MethodHandle adaptSystemCall(MethodHandle target,
+                                               String stateName) {
+        // Invariants checked in the BasicKey record
+        final BasicKey basicKey = new BasicKey(target, stateName);
+
+        // ((int | long), MemorySegment)(int | long)
+        final MethodHandle basicHandle = BASIC_HANDLE_CACHE.apply(basicKey);
+
+        // Make `target` specific adaptations of the basic handle
+
+        // Pre-pend all the parameters from the `target` MH.
+        // (C0=MemorySegment, C1-Cn, MemorySegment)(int|long)
+        MethodHandle innerAdapted = MethodHandles.collectArguments(basicHandle, 0, target);
+
+        final int[] perm = new int[target.type().parameterCount() + 1];
+        for (int i = 0; i < target.type().parameterCount(); i++) {
+            perm[i] = i;
+        }
+        // Last takes first
+        perm[target.type().parameterCount()] = 0;
```

Comment:
```suggestion
        for (int i = 0; i < perm.length - 1; i++) {
            perm[i] = i;
        }
        // Last takes first
        perm[perm.length - 1] = 0;
```
Here you can use perm.length - 1